File: ios/TodayInSpaceHistory/DataModels/ImageProviderService.swift
//
//  ImageProviderService.swift
//  TodayInSpaceHistory
//
//  Created by Mikołaj Linczewski on 09/08/2026.
//

import Foundation

protocol ImageProviderServiceProtocol {
    func loadTodaysImage() async throws -> (item: Item, imageURLs: [String])
}

final class ImageProviderService: ImageProviderServiceProtocol {
    private let client: NetworkClientProtocol
    
    init(client: NetworkClientProtocol = NetworkClient()) {
        self.client = client
    }
    
    func loadTodaysImage() async throws -> (item: Item, imageURLs: [String]) {
        let response: APIResponse = try await client.makeRequest(endpoint: .search)
        let items = response.collection.items ?? []
        let todaysItems = items.filter(\.matchesTodaysAnniversary)
        guard let item = (todaysItems.isEmpty ? items : todaysItems).randomElement() else {
            throw Errors.ImageProvider.noItems
        }
        guard let href = item.href else {
            throw Errors.ImageProvider.missingAssetURL
        }
        let rawURLs: [String] = try await client.fetchImages(url: href)
        let imageURLs = rawURLs.map(\.rewritingHTTPSchemeToHTTPS)
        return (item, imageURLs)
    }
}

extension String {
    /// Rewrites only the `http://` scheme prefix; leaves `https://` unchanged.
    var rewritingHTTPSchemeToHTTPS: String {
        guard lowercased().hasPrefix("http://") else { return self }
        return "https://" + dropFirst("http://".count)
    }
}

File: android/app/src/main/java/miko/todayinspacehistory/data/ImageProviderService.kt
package miko.todayinspacehistory.data

import miko.todayinspacehistory.data.model.Item
import miko.todayinspacehistory.data.network.NetworkClient
import miko.todayinspacehistory.data.network.NetworkClientImpl
import miko.todayinspacehistory.util.rewritingHTTPSchemeToHTTPS

data class TodaysImage(
    val item: Item,
    val imageUrls: List<String>,
)

interface ImageProviderService {
    suspend fun loadTodaysImage(): TodaysImage
}

class ImageProviderServiceImpl(
    private val client: NetworkClient = NetworkClientImpl(),
) : ImageProviderService {

    override suspend fun loadTodaysImage(): TodaysImage {
        val response = client.search()
        val items = response.collection.items.orEmpty()
        val todaysItems = items.filter { it.matchesTodaysAnniversary }
        val pool = if (todaysItems.isEmpty()) items else todaysItems
        val item = pool.randomOrNull()
            ?: throw Errors.ImageProvider.NoItems
        val href = item.href
            ?: throw Errors.ImageProvider.MissingAssetURL
        val rawUrls = client.fetchImageUrls(href)
        val imageUrls = rawUrls.map { it.rewritingHTTPSchemeToHTTPS }
        return TodaysImage(item = item, imageUrls = imageUrls)
    }
}
